feat(db): add SQLiteMigrator to perform db migrations

psychedelicious opened the pull request:
## What type of PR is this? (check all applicable)

- [x] Refactor
- [x] Feature
- [ ] Bug Fix
- [x] Optimization
- [ ] Documentation Update
- [ ] Community Node Submission

## Have you discussed this change with the InvokeAI team?

- [x] Yes
- [ ] No, because:

## Description

This PR enhances our SQLite database with migration logic.

### `SQLiteMigrator` class

The new `SQLiteMigrator` class handles safely running database migrations. It is initialized in the `SqliteDatabase` class's init, and immediately runs all database migrations.

### `Migration` class

Migrations are reprsented by a `Migration` class, which has 3 attributes:

- `db_version: int`: The database version this migration results in.
- `app_version: str`: The semver app version this migration is run for.
- `migrate: Callable[[sqlite3.Cursor], None]`: A function that performs the migration. It receives a cursor _only_, but can do anything it wants to do. A convention is established for these functions.

All schema-creating SQL now lives in a `migrate` function. We haven't needed to make any data migrations yet, but when we do, this will also be handled within one of these callbacks.

### Migration Flow

First, migrations are registered with `SQLiteMigrator` with it's `register_migration` method. This performs some basic checks of the migration version.

After registering all migrations, they are run with the `run_migrations` method. This does a few things:

- Creates a `version` table in the DB, if it doesn't already exist. This table has `db_version INTEGER`, `app_version TEXT` and `migrated_at DATETIME` columns.
- Sort the migrations by their `db_version`.
- Do some checks to see if we need a migration.
- Backs up the database (if it's a file database). The migration bails out if this fails.
- Runs each migration. If there is a problem, restore from backup.

### Included Migrations

Migrations are in `invokeai/app/services/shared/sqlite/migrations`.

#### `migrate_1.py`

All\* schema SQL up to 3.4.0post2 is in `migration_1.py`. Running only this migration should result in a database that is identical to the one you get from starting up 3.4.0post2.

SQL in this migration is **idempotent** (same as it was when the SQL was spread across the various services).

#### `migrate_2.py`

Schema changes through 3.5.0 (the upcoming release) are in `migration_2.py`.

SQL in this migration is **not idempotent**. Future migrations need not be idempotent, as the migration logic ensures each will only be run once.

### \*Caveat - ItemStorage

This class provides a generic document-db-like interface for storing objects. Our `graph_executions` and `graphs` tables are created and managed by this service. This PR does not touch this class and therefore does not touch either of those two tables.

We can decide how to handle those tables in the future as the need arises.

### Change to Model Manager Metadata table

I noticed that there is a `model_manager_metadata` table which included the app version, and whose `version` property wasn't accessed outside the service.

I believe the new `version` table fulfills the purpose of this table, and have removed it.

@lstein Please let me know if this is not right.

## QA Instructions, Screenshots, Recordings

1.  Case 1 - Upgrade

    - Back up your 3.4.0post2 database
    - Run this PR
    - It should upgrade your database and everything should work exactly like it did before

2.  Case 2 - New Install

    - Move your database out of the invoke root so that when the app starts, it creates a new one
    - Run this PR
    - It should work just like a new install

3.  Case 3 - With an In-Memory Database

    - Enable the in-memory memory database (set `use_memory_db` under `Paths` in `invokeai.yaml` to `true`)
    - Run this PR
    - It should work just like a new install


## Added/updated tests?

- [x] Yes: Fairly comprehensive tests are added for the `SQLiteMigrator`.
- [ ] No : _please replace this line with details on why tests
      have not been included_

psychedelicious commented:
Oh, fun, the tempdirectory is failing github actions on windows :(

lstein commented:
This looks neat. I'll start reviewing tomorrow.

lstein commented on invokeai/app/services/shared/sqlite/migrations/migration_2.py:
```diff
@@ -0,0 +1,97 @@
+    """Adds the `workflow_library` table and drops the `workflows` and `workflow_images` tables."""
+    tables = [
+        """--sql
+        CREATE TABLE workflow_library (
```
I wonder if this and similar methods should use `CREATE TABLE IF NOT EXISTS` instead of the unconditional `CREATE TABLE`. This would future proof against a crash at a critical moment when some of the tables have been added and others aren't.

psychedelicious replied:
I've gone ahead and done this for this particular migration, but I've improved the resiliency of the migrator and don't think we need this kind of idempotency for migrations - we rely on not running a migration multiple times for that.

lstein commented on invokeai/app/services/shared/sqlite/sqlite_migrator.py:
```diff
@@ -0,0 +1,210 @@
+                migration.migrate(self._cursor)
+                # Migration callbacks only get a cursor; they cannot commit the transaction.
+                self._conn.commit()
+                self._set_version(db_version=migration.db_version, app_version=migration.app_version)
```
In the unlikely but conceivable case that there is a crash/power outage between the `commit()` in the line above and the `_set_version()` here you could end up with a database with the newer version's tables but the older  version number.  When the app is restarted the migration system will re-run the migration and then fail because the new tables and indexes are already there. I think `_set_version()` should be moved above the `commit()` to avoid this. The call to `commit()` in `_set_version()` should then be removed.

lstein commented on invokeai/app/services/shared/sqlite/sqlite_database.py:
```diff
@@ -30,15 +33,20 @@ def __init__(self, config: InvokeAIAppConfig, logger: Logger):
+        migrator = SQLiteMigrator(conn=self.conn, database=self.database, lock=self.lock, logger=self._logger)
+        migrator.register_migration(migration_1)
+        migrator.register_migration(migration_2)
+        migrator.run_migrations()
```
I'd suggest writing out an empty file at the beginning of the migration process called (something like) `.migration_in_progress` and then removing it after the migration is done. In the event of a crash, this file will still be there. You can then detect this eventuality and restore the database from the backup  you created at the beginning of the migration.

psychedelicious replied:
Good idea!

lstein reviewed (commented):
I've reviewed and tested. Everything looks good, but I made some minor suggestions to increase robustness against crashes (keyboard interrupt, power fault) during the migration process.

lstein commented on invokeai/app/services/shared/sqlite/sqlite_migrator.py:
```diff
@@ -0,0 +1,210 @@
+                except MigrationError:
+                    if backup_path is not None:
+                        self._logger.error(f" Restoring from {backup_path}")
+                        self._restore_db(backup_path)
```
The alternative way to guard against failed migrations is to make a copy of the database as before, make the schema changes _in the new database_, and then when everything completes move the old database away (thereby becoming the backup) and renaming the copy to become the current database.

psychedelicious replied:
Oh, that sounds much safer

RyanJDick commented on invokeai/app/services/shared/sqlite/sqlite_database.py:
```diff
@@ -30,15 +33,20 @@ def __init__(self, config: InvokeAIAppConfig, logger: Logger):
 
         self.conn.execute("PRAGMA foreign_keys = ON;")
 
+        migrator = SQLiteMigrator(conn=self.conn, database=self.database, lock=self.lock, logger=self._logger)
```
I'd be inclined to move the migration logic out of `SqliteDatabase`. I think of `SqliteDatabase` more as a DB connection, and I think it's more natural to tie the migration to app startup rather than `SqliteDatabase` construction.

RyanJDick commented on invokeai/app/services/shared/sqlite/sqlite_migrator.py:
```diff
@@ -0,0 +1,210 @@
+        """Creates a version table for the database, if one does not already exist."""
+        with self._lock:
+            try:
+                self._cursor.execute("SELECT name FROM sqlite_master WHERE type='table' AND name='version';")
+                if self._cursor.fetchone() is not None:
+                    return
```
Doesn't `CREATE TABLE IF NOT EXISTS` achieve the same thing?

psychedelicious replied:
It does, but:
- a few lines down we also insert the initial `0` version into this table
- and then write a log saying we created the table

This check lets us return early and skip that logic

RyanJDick commented on invokeai/app/services/shared/sqlite/sqlite_migrator.py:
```diff
@@ -0,0 +1,210 @@
+                    return
+                self._cursor.execute(
+                    """--sql
+                    CREATE TABLE IF NOT EXISTS version (
```
nit: `migrations` might better describe what this table contains.

psychedelicious replied:
Agreed

psychedelicious commented:
@lstein Thanks for reviewing!

I thought a bit more about this since I created the PR and think my design has a major flaw.

Migrations currently just have a resultant `db_version`, and execution is based on *inequality* (less-than/greater-than) comparison between version numbers. There is an implicit assumption there about db version numbers which may be false.

Instead, the execution should be based on *equality* comparison. Migrations should have should have a `from_db_version` in addition to the `to_db_version`, and they only run if `current_db_version == from_db_version`.

Does my thinking here make sense? If so, I'll incorporate this and your other feedback and revise this PR

RyanJDick commented on invokeai/app/services/shared/sqlite/migrations/migration_1.py:
```diff
@@ -0,0 +1,370 @@
+        cursor.execute(stmt)
+
+
+migration_1 = Migration(db_version=1, app_version="3.4.0", migrate=_migrate)
```
What's the reason for storing the app_version? I'm wondering:
1) if it's needed at all, and
2) whether we should instead store the version of the app that _ran_ the migration in the DB

psychedelicious replied:
Included because @lstein had the app_version in the `model_manager_metadata` table and I made an assumption that there was a need for it.

Also figured it was useful for to know during which app release the db was migrated.

I commented just before your review about a flaw in the design of this: https://github.com/invoke-ai/InvokeAI/pull/5227#issuecomment-1846261023

If we store the app version at all, I think we should store both the "from" and "to" versions. We'd need both of those data to derive meaningful information. For example, if a user installing 3.5.0 from 3.2.0 is different than from 3.4.0, and knowing the version they came from is useful for troubleshooting.

psychedelicious replied:
I've removed this as I don't think it is useful

psychedelicious commented on invokeai/app/api/dependencies.py:
```diff
@@ -66,9 +72,25 @@ def initialize(config: InvokeAIAppConfig, event_handler_id: int, logger: Logger
+            logger=logger,
+            log_sql=config.log_sql,
+        )
+        migration_2.register_post_callback(partial(migrate_embedded_workflows, logger=logger, image_files=image_files))
```
This post-migration callback populates the `has_workflow` column for the `images` table. It needs a logger and image file service to run.

I opted to provide those dependencies using a partial for two reasons:
1. It reduces coupling of `SQLiteMigrator` and the rest of the app.
2. It is more flexible for future migrations, which may need unique dependencies.

psychedelicious commented on invokeai/app/api/dependencies.py:
```diff
@@ -66,9 +72,25 @@ def initialize(config: InvokeAIAppConfig, event_handler_id: int, logger: Logger
+        migrator.register_migration(migration_2)
+        did_migrate = migrator.run_migrations()
+
+        if not db.is_memory and did_migrate:
+            db.reinitialize()
```
We need to re-connect to the database after doing the migrations, but only if 1) there was a migration and 2) we are using a file database.

This closes the `SqliteDatabase`'s connection and runs its `__init__` logic again.

If we do this with a memory db, it wipes the db.

RyanJDick replied:
This explanation was helpful. Can you add it as an in-code comment?

psychedelicious replied:
Superseded

psychedelicious commented on invokeai/app/services/shared/sqlite/migrations/migration_2_post.py:
This is the only data migration needed so far, and it has additional dependencies. See docstring and PR comment in `dependencies.py`.

psychedelicious commented on invokeai/app/services/shared/sqlite/sqlite_migrator.py:
```diff
@@ -0,0 +1,307 @@
+    the dependencies and register the partial as the migration callback.
+    """
+
+    from_version: int = Field(ge=0, strict=True, description="The database version on which this migration may be run")
+    to_version: int = Field(ge=1, strict=True, description="The database version that results from this migration")
```
Migrations are now aware of what version they can migrate from and to. This makes it much, much harder to bork.

psychedelicious commented on invokeai/app/services/shared/sqlite/sqlite_migrator.py:
```diff
@@ -0,0 +1,312 @@
+        return sorted(self._migrations, key=lambda m: m.to_version)[-1].to_version
+
+
+def get_temp_db_path(original_db_path: Path) -> Path:
+    """Gets the path to the temp database."""
+    return original_db_path.parent / original_db_path.name.replace(".db", ".db.temp")
```
This is the copy that we do the migration on

psychedelicious commented on invokeai/app/services/shared/sqlite/sqlite_migrator.py:
```diff
@@ -0,0 +1,312 @@
+    return original_db_path.parent / original_db_path.name.replace(".db", ".db.temp")
+
+
+def get_backup_db_path(original_db_path: Path) -> Path:
+    """Gets the path to the final backup database."""
+    timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
+    return original_db_path.parent / f"{original_db_path.stem}_backup_{timestamp}.db"
```
This is where the original database is backed up to before we move the migrated copy back into its place

psychedelicious commented on invokeai/app/services/shared/sqlite/sqlite_migrator.py:
```diff
@@ -0,0 +1,313 @@
+                self._logger.debug(
+                    f"Successfully migrated database from {migration.from_version} to {migration.to_version}"
+                )
+            except Exception as e:
```
This is left purposefully wide - I think we want to catch everything and handle in the same way.

psychedelicious commented on tests/test_sqlite_migrator.py:
I haven't updated the tests since my recent round of changes. These are expected to fail.

psychedelicious replied:
I've updated the tests for the revised migrator implementation and all are now passing

psychedelicious replied:
I neglected to check the other tests, which need the migrator in their fixtures. The tests & fixtures are getting really message but fixing that is not in scope here

psychedelicious commented:
@RyanJDick @lstein 

I've incorporated the your feedback, made some improvements and added a data migration for the workflow library:
- Move the migrator and its calls out of the `SqliteDatabase` class
- Use the safer backup strategy suggested by @lstein, where the migration is run on a copy of the database. If successful, the original database is renamed and retained as a backup, and the copy renamed to the original database's name.
  - We get free catastrophic failure detection from this. If a copy exists when the app starts up, it means the previous migration was interrupted. We just try again in this case, logging a warning.
- Revise the `Migration` class to use `from_version` and `to_version`. This gives us much more confidence that we never run the wrong migrations.
- Clean up the handling when using an in-memory db - we don't need to make backups and such
- Add support for pre- and post- migration callbacks
- Add the workflow library data migration as a post-migration callback. This data migration populates the new `has_workflow` column in the db. It needs access to a logger and image file service. These dependencies are provided by making a partial.
- More comments and cleanup

Appreciate another round of reviews here.

psychedelicious commented on invokeai/app/services/shared/sqlite/sqlite_migrator.py:
```diff
@@ -0,0 +1,317 @@
+    def __init__(
+        self,
+        db_path: Path | None,
+        conn: sqlite3.Connection,
```
To support in-memory databases, we need to share a connection for the lifetime of the application, so this must be passed in.

RyanJDick commented on invokeai/app/services/shared/sqlite_migrator/sqlite_migrator_impl.py:
```diff
@@ -0,0 +1,212 @@
+                    temp_db_path=temp_db_path,
+                    original_db_path=self._db_path,
+                )
+                self._logger.info(f"Migration successful. Original DB backed up to {backup_db_path}")
```
SQLite is ACID-compliant and has [very robust failure handling](https://www.sqlite.org/transactional.html).

I'm questioning if our file shuffling is really adding any significant protection, because it definitely has some inconveniences:
- Backup DB files will accumulate as we run more migrations (unless I've missed the cleanup code somewhere?).
- If there is a failure during the final file renaming step, manual action could be required to recover (albeit unlikely)

psychedelicious replied:
lol what a docs page. Handling files has been a pain, requiring me to boot up windows *hurk* to fix tests. I trust SQLite far more than myself, manual shuffling files. I'm happy to rely on SQLite.

psychedelicious replied:
I've removed all file stuff from the migrator. I have trust in SQLite.

RyanJDick commented on invokeai/app/services/shared/sqlite_migrator/sqlite_migrator_common.py:
```diff
@@ -0,0 +1,109 @@
+        # register() ensures that there is only one migration with a given from_version, so this is safe.
+        return next((m for m in self._migrations if m.from_version == from_version), None)
+
+    def validate_migration_chain(self) -> None:
+        """
+        Validates that the migrations form a single chain of migrations from version 0 to the latest version.
+        Raises a MigrationError if there is a problem.
+        """
```
Related to my other comment about the motivation behind pre/post migration callbacks: validation doesn't currently account for them - this seems error-prone.

psychedelicious replied:
A migrate callback does not change the version (or it shouldn't, at least) - the migrator itself does that. This should be safe - am I missing something?

psychedelicious replied:
Superseded

psychedelicious commented on invokeai/app/api/dependencies.py:
```diff
@@ -67,9 +73,25 @@ def initialize(config: InvokeAIAppConfig, event_handler_id: int, logger: Logger
         db = SqliteDatabase(config, logger)
 
+        migrator = SQLiteMigrator(
+            db_path=db.database if isinstance(db.database, Path) else None,
+            conn=db.conn,
+            lock=db.lock,
```
I had it like that initially but it made testing a bit more involved, because we need to provide the `config` and `logger` in the fixes. I removed it early in the last iteration of changes.

In hindsight, we are instantiating `SqliteDatabase` with a `config` and `logger` in other tests, so I think it's fine to pass in the whole object. It also lets us do the pesky `reinitialize` step within the migration logic, reducing surface area for problems with that.

psychedelicious replied:
I've simplified `SqliteDatabase` to not be dependent on a particular `config` object, instead it gets only what it needs: a path, a logger, and a verbose flag.

This makes `SqliteMigrator` simpler, too. Thanks for the nudge

psychedelicious commented on invokeai/app/services/shared/sqlite_migrator/sqlite_migrator_common.py:
```diff
@@ -0,0 +1,109 @@
+
+    @model_validator(mode="after")
+    def validate_to_version(self) -> "Migration":
+        if self.to_version <= self.from_version:
+            raise ValueError("to_version must be greater than from_version")
```
I anticipated cases like that but I suppose we have full control over this and could enforce sequential db versions.

psychedelicious replied:
Implemented

psychedelicious commented on invokeai/app/services/shared/sqlite_migrator/sqlite_migrator_common.py:
```diff
@@ -0,0 +1,109 @@
+    )
+    post_migrate: list[MigrateCallback] = Field(
+        default=[], description="A list of callbacks to run after the migration"
+    )
```
I've implemented what I think is a much nicer system for the migrations. I'll mark this resolved and add a top-level comment.

lstein commented:
> @lstein Thanks for reviewing!
> 
> I thought a bit more about this since I created the PR and think my design has a major flaw.
> 
> Migrations currently just have a resultant `db_version`, and execution is based on _inequality_ comparison between version numbers, e.g. `current_db_version < migration.db_version`. There is an implicit assumption there about db version numbers which may be false.
> 
> Instead, the execution should be based on _equality_ comparison. Migrations should have should have a `from_db_version` in addition to the `to_db_version`, and they only run if `current_db_version == migration.from_db_version`.
> 
> Does my thinking here make sense? If so, I'll incorporate this and your other feedback and revise this PR

Yes, I think this is the right test to perform.

psychedelicious commented:
@RyanJDick @lstein Thanks for your reviews. I think I'm in final draft territory now.

### SQLite transactions > backups
As @RyanJDick [pointed out](https://github.com/invoke-ai/InvokeAI/pull/5227#discussion_r1423080099), SQLite has excellent safety around transactions and the backing up the database to a file is largely unnecessary. 

I've removed the logic around this, which makes the migrator a lot cleaner. However, I wonder if it makes sense to back up the database before releases anyways - what do you think? This could be outside the migrator, or a very simple function within the migrator.

### Migration dependencies
The v3.5.0 migration requires access to the image files service to check if they have an embedded workflow. I had handled this by manually creating `partial`s of the migrate callbacks and handling them as "post" callbacks, but this was messy and error-prone.

I've revised this with a more resilient (and hopefully elegant) design. Migrations may declare dependencies at instantiation, which then must be provided at runtime. This enabled the 3.5.0 migration to use the image files service after it has been initialized.

I've added documentation here:
https://github.com/invoke-ai/InvokeAI/pull/5227/files#diff-cf84e4dd7ecaa30bf7862e256fe4723f7d7ec0b9b362d65d41cb54cf85da872aR51-R106

Hopefully this just needs one more round of reviews.

psychedelicious commented on invokeai/app/services/shared/sqlite_migrator/migrations/migration_2.py:
```diff
@@ -0,0 +1,196 @@
+    cursor.execute("DROP TABLE IF EXISTS model_manager_metadata;")
+
+
+def _recreate_model_config(cursor: sqlite3.Cursor) -> None:
+    """
+    Drops the `model_config` table, recreating it.
+
+    In 3.4.0, this table used explicit columns but was changed to use json_extract 3.5.0.
+
+    Because this table is not used in production, we are able to simply drop it and recreate it.
+    """
+
+    cursor.execute("DROP TABLE IF EXISTS model_config;")
+
+    cursor.execute(
+        """--sql
+        CREATE TABLE IF NOT EXISTS model_config (
+            id TEXT NOT NULL PRIMARY KEY,
+            -- The next 3 fields are enums in python, unrestricted string here
+            base TEXT GENERATED ALWAYS as (json_extract(config, '$.base')) VIRTUAL NOT NULL,
+            type TEXT GENERATED ALWAYS as (json_extract(config, '$.type')) VIRTUAL NOT NULL,
+            name TEXT GENERATED ALWAYS as (json_extract(config, '$.name')) VIRTUAL NOT NULL,
+            path TEXT GENERATED ALWAYS as (json_extract(config, '$.path')) VIRTUAL NOT NULL,
+            format TEXT GENERATED ALWAYS as (json_extract(config, '$.format')) VIRTUAL NOT NULL,
+            original_hash TEXT, -- could be null
+            -- Serialized JSON representation of the whole config object,
+            -- which will contain additional fields from subclasses
+            config TEXT NOT NULL,
+            created_at DATETIME NOT NULL DEFAULT(STRFTIME('%Y-%m-%d %H:%M:%f', 'NOW')),
+            -- Updated via trigger
+            updated_at DATETIME NOT NULL DEFAULT(STRFTIME('%Y-%m-%d %H:%M:%f', 'NOW')),
+            -- unique constraint on combo of name, base and type
+            UNIQUE(name, base, type)
+        );
+        """
+    )
```
I've updated this table to match these changes:
https://github.com/invoke-ai/InvokeAI/pull/5264#discussion_r1423406713

This migration represents changes from 3.4.0 to 3.5.0, so we can continue to update this migration file until v3.5.0 is released. We'll need a new migration file for subsequent changes

RyanJDick commented on invokeai/app/services/shared/sqlite_migrator/sqlite_migrator_common.py:
````diff
@@ -0,0 +1,209 @@
+    before the migration is run. The migrator provides dependencies to the migrate callback,
+    raising an error if a dependency is missing or was provided the wrong type.
+
+    Example Usage:
+    ```py
+    # Create a migration dependency. This migration needs access the image files service, so we set the type to the ABC of that service.
+    image_files_dependency = MigrationDependency(name="image_files", dependency_type=ImageFileStorageBase)
+
+    # Define the migrate callback. The dependency may be accessed by name in the kwargs. The migrator will ensure that the dependency is of the required type.
+    def migrate_callback(cursor: sqlite3.Cursor, **kwargs) -> None:
+        image_files = kwargs[image_files_dependency.name]
+        # Do something with image_files
+        ...
+
+    # Instantiate the migration, including the dependency.
+    migration = Migration(
+        from_version=0,
+        to_version=1,
+        migrate_callback=migrate_callback,
+        dependencies={image_files_dependency.name: image_files_dependency},
+    )
+
+    # Provide the dependency before registering the migration.
+    # (DiskImageFileStorage is an implementation of ImageFileStorageBase)
+    migration.provide_dependency(name="image_files", value=DiskImageFileStorage())
````
Here's an alternative solution to the migration dependency problem. Take it or leave it. Your implementation is already in good shape (and you've spent more time thinking about this than me), so I'll defer to you to make the call. But, I'll give a couple reasons why I think this is better.

```python
# ----- migrate_2.py -----
class Migration2Function:
    def __init__(self, image_files: ImageFileStorageBase, logger: Logger):
        self._image_files = image_files
        self._logger = logger

    def __call__(self, cursor: sqlite3.Cursor):
        # Run the migration with access to the self._image_files and self._logger dependencies.
        ...

def build_migration_2(image_files: ImageFileStorageBase, logger: Logger) -> Migration:
    # Could alternatively use `partial(...)` here if you prefer that over using a class.
    migration_fn = Migration2Function(image_files, logger)

    return Migration(
        from_version=1,
        to_version=2,
        migrate_callback=migration_fn,
    )

# ----- invokeai/app/api/dependencies.py -----
...

migrator = SQLiteMigrator(db=db)
migrator.register_migration(build_migration_1())
migrator.register_migration(build_migration_2(image_files, logger))
migrator.run_migrations()
...

```

Advantages of this approach:
- Simpler. Can drop all of the`MigrationDependency` handling.
- We don't have to worry about runtime dependency type checking. We benefit from static type checking (or at least as much as we ever do in python).
- No code gets run on import of the migration files. I.e. when we import migration_1.py, we don't instantiate a migration, we just declare classes/functions. I try to avoid running code on import, because it can lead to confusing bugs where import order matters.

psychedelicious replied:
Oh, this is much nicer. I was conceptually orbiting something like this, but ended up with the contrived dependency handling. I can't think of any reason to use my implementation of the dependencies. I'll do this instead, thanks!

Using a class also allows the smaller logical migration functions to be local to the class - having those in the module scope was bothering me.

Also cool that you can just make a class callable with `__call__`. I knew you could make protocols like this but it never clicked.

RyanJDick commented on tests/fixtures/sqlite_database.py:
```diff
@@ -0,0 +1,33 @@
+
+
+@pytest.fixture
+def create_sqlite_database() -> CreateSqliteDatabaseFunction:
```
Why pass this around as a fixture rather than just defining a function and importing it?

psychedelicious replied:
Good question. I saw some examples of using pytest fixtures in this way, but they were using the pattern to compose fixtures. I'm apparently not composing anything here, so the pattern is only adding confusion. I'll remove it.

RyanJDick reviewed (commented):
I'm very pleased with the evolution of this PR - it's looking great. Just a couple final comments from me.

psychedelicious commented:
@RyanJDick
I've implemented your suggestion for handling dependencies - great idea. Comments updated and the dependency tests are removed as there's nothing to test.

I made some other small changes:
- Minor improvement to db initialization. That is now in an `init_db` function, which is called both by the app and test suite. This ensures the we keep the tests and app using the same set of migrations. A mock is used for the image files dependency.
- I renamed "SQLiteMigrator" to "SqliteMigrator" to be consistent with the rest of the app
- Renamed `Migration.migrate_callback` to simply `callback`
- Updated some comments

This this is ready

RyanJDick reviewed (approved):
👌

hipsterusername commented:
Let it be done.
